src/cmd/groff/contrib/pdfmark/pdfroff.man
.TH PDFROFF @MAN1EXT@ "@MDATE@" "Groff Version @VERSION@"
.\" --------------------------------------------------------------------
.\" Legal Matters
.\" --------------------------------------------------------------------
.ig
pdfroff.1

File position: <groff-source>/contrib/pdfmark/pdfroff.man

Last update: 
30 Jan 2007.

This file is part of groff, the GNU roff type-setting system.

Copyright (C) 2005, 2006, 2007 Free Software Foundation, Inc.
written by Keith Marshall <keith.d.marshall@ntlworld.com>

Permission is granted to copy, distribute and/or modify this document
under the terms of the GNU Free Documentation License, Version 1.1 or
any later version published by the Free Software Foundation; with no
Front-Cover Texts, no Back-Cover Texts, and the following Invariant
Sections:--

    a)  This "Legal Matters" section, extending from the start of
        the document, to the end of the enclosing ".ig" section.

    b)  The entire section bearing the heading "AUTHOR", extending
        from the ".SH AUTHOR" tag, to the end of the document.

A copy of the Free Documentation License is included as a file called
FDL in the main directory of the groff source package.
..
.
.
.\" --------------------------------------------------------------------
.\" Local macro definitions
.
.hw pdfmark
.
.de NH
.  hy 0
\&\\$*
.  hy
..
.
.
.\" --------------------------------------------------------------------
.
.SH NAME
.
pdfroff \- create PDF documents using groff
.
.
.\" --------------------------------------------------------------------
.
.SH SYNOPSIS
.
.SY pdfroff
.OP \-abcegilpstzCEGNRSUVXZ
.OP \-d cs
.OP \-f fam
.OP \-F dir
.OP \-I dir
.OP \-L arg
.OP \-m name
.OP \-M dir
.OP \-n num
.OP \-o list
.OP \-P arg
.OP \-r cn
.OP \-T dev
.OP \-w name
.OP \-W name
.OP \-\-emit\-ps
.OP \-\-no\-toc\-relocation
.OP \-\-no-kill\-null\-pages
.OP \-\-stylesheet=\fIname\fP
.OP \-\-no\-pdf\-output
.OP \-\-pdf\-output=\fIname\fP
.OP \-\-no\-reference\-dictionary
.OP \-\-reference\-dictionary=\fIname\fP
.OP \-\-report\-progress
.OP \-\-keep\-temporary\-files
.I file .\|.\|.
.
.SY pdfroff
.B \-h
|
.B \-\-help
.
.SY pdfroff
.B \-v
|
.B \-\-version
.RI [ option\ .\|.\|. ]
.YS
.
.
.\" --------------------------------------------------------------------
.
.SH DESCRIPTION
.
.B pdfroff
is a wrapper program for the GNU text processing system,
.BR  groff .
It transparently handles the mechanics of multiple pass
.B groff
processing, when applied to suitably marked up
.B groff
source files,
such that tables of contents and body text are formatted separately,
and are subsequently combined in the correct order, for final publication
as a single PDF document.
A further optional
\*(lqstyle sheet\*(rq
capability is provided;
this allows for the definition of content which is required to precede the
table of contents, in the published document.
.
.P
For each invocation of
.BR pdfroff ,
the ultimate
.B groff
output stream is post-processed by the GhostScript interpreter,
to produce a finished PDF document.
.
.P
.B pdfroff
makes no assumptions about, and imposes no restrictions on,
the use of any
.B groff
macro packages which the user may choose to employ,
in order to achieve a desired document format;
however, it
.I does
include specific built in support for the
.B pdfmark
macro package, should the user choose to employ it.
Specifically, if the
.I pdfhref
macro, defined in the
.B pdfmark.tmac
package, is used to define public reference marks,
or dynamic links to such reference marks, then
.B pdfroff
performs as many preformatting
.B groff
passes as required, up to a maximum limit of
.IR four ,
in order to compile a document reference dictionary,
to resolve references, and to expand the dynamically defined
content of links.
.
.
.\" --------------------------------------------------------------------
.
.SH USAGE
.
The command line is parsed in accordance with normal GNU conventions,
but with one exception \(em when specifying any short form option
(i.e., a single character option introduced by a single hyphen),
and if that option expects an argument, then it
.I must
be specified independently (i.e., it may
.I not
be appended to any group of other single character short form options).
.
.P
Long form option names (i.e., those introduced by a double hyphen)
may be abbreviated to their minimum length unambiguous initial
substring.
.
.P
Otherwise,
.B pdfroff
usage closely mirrors that of
.B groff
itself.
Indeed,
with the exception of the
.BR \-h ,
.BR \-v ,
and
.BI \-T \ dev
short form options, and
all long form options,
which are parsed internally by
.BR pdfroff ,
all options and file name arguments specified on the command line are
passed on to
.BR groff ,
to control the formatting of the PDF document.
Consequently,
.B pdfroff
accepts all options and arguments, as specified in
.BR groff (@MAN1EXT@),
which may also be considered as the definitive reference for all standard
.BR pdfroff
options and argument usage.
.
.
.\" --------------------------------------------------------------------
.
.SH OPTIONS
.
.B pdfroff
accepts all of the short form options
(i.e., those introduced by a single hyphen),
which are available with
.B groff
itself.
In most cases, these are simply passed transparently to
.BR groff ;
the following, however, are handled specially by
.BR pdfroff .
.
.TP
.B \-h
Same as
.BR \-\-help ;
see below.
.
.TP
.B \-i
Process standard input, after all other specified input files.
This is passed transparently to
.BR groff ,
but, if grouped with other options, it
.I must
be the first in the group.
Hiding it within a group
breaks standard input processing, in the multiple pass
.B groff
processing context of
.BR pdfroff .
.
.TP
.BI \-T \ dev
Only
.B \-T\ ps
is supported by
.BR pdfroff .
Attempting to specify any other device causes
.B pdfroff
to abort.
.
.TP
.B \-v
Same as
.BR \-\-version ;
see below.
.
.P
See
.BR groff (@MAN1EXT@)
for a description of all other short form options,
which are transparently passed through
.BR pdfroff
to
.BR groff .
.
.P
All long form options
(i.e., those introduced by a double hyphen)
are interpreted locally by
.BR pdfroff ;
they are
.B not
passed on to
.BR groff ,
unless otherwise stated below.
.
.TP
.B \-\-help
Causes
.B pdfroff
to display a summary of the its usage syntax, and supported options,
and then exit.
.
.TP
.B \-\-emit\-ps
Suppresses the final output conversion step,
causing
.B pdfroff
to emit PostScript output instead of PDF.
This may be useful,
to capture intermediate PostScript output,
when using a specialised postprocessor,
such as
.I gpresent
for example,
in place of the default
.I GhostScript
PDF writer.
.
.TP
.B \-\-keep\-temporary\-files
Suppresses the deletion of temporary files,
which normally occurs after
.B pdfroff
has completed PDF document formatting;
this may be useful,
when debugging formatting problems.
.IP
See section
.BR FILES ,
for a description of the temporary files used by
.BR pdfroff .
.
.TP
.B \-\-no\-pdf\-output
May be used with the
.BI \%\-\-reference\-dictionary= name
option (described below) to eliminate the overhead of PDF formatting,
when running
.B pdfroff
to create a reference dictionary, for use in a different document.
.
.TP
.B \-\-no\-reference\-dictionary
May be used to eliminate the overhead of creating a reference dictionary,
when it is known that the target PDF document contains no public
references, created by the
.I pdfhref
macro.
.
.TP
.B \-\-no\-toc\-relocation
May be used to eliminate the extra
.B groff
processing pass,
which is required to generate a table of contents,
and relocate it to the start of the PDF document,
when processing any document which lacks an automatically
generated table of contents.
.
.TP
.B \-\-no\-kill\-null\-pages
While preparing for simulation of the manual collation step,
which is traditionally required to relocate of a
.I "table of contents"
to the start of a document,
.B pdfroff
accumulates a number of empty page descriptions
into the intermediate
.I PostScript
output stream.
During the final collation step,
these empty pages are normally discarded from the finished document;
this option forces
.B pdfroff
to leave them in place.
.
.TP
.BI \-\-pdf\-output= name
Specifies the name to be used for the resultant PDF document;
if unspecified, the PDF output is written to standard output.
A future version of
.B pdfroff
may use this option,
to encode the document name in a generated reference dictionary.
.
.TP
.BI \-\-reference\-dictionary= name
Specifies the name to be used for the generated reference dictionary file;
if unspecified, the reference dictionary is created in a temporary file,
which is deleted when
.B pdfroff
completes processing of the current document.
This option
.I must
be specified, if it is desired to save the reference dictionary,
for use in references placed in other PDF documents.
.
.TP
.B \-\-report\-progress
Causes
.B pdfroff
to display an informational message on standard error,
at the start of each
.B groff
processing pass.
.
.TP
.BI \-\-stylesheet= name
Specifies the name of an
.IR "input file" ,
to be used as a style sheet for formatting of content,
which is to be placed
.I before
the table of contents,
in the formatted PDF document.
.
.TP
.B \-\-version
Causes
.B pdfroff
to display a version identification message.
The entire command line is then passed transparently to
.BR groff ,
in a
.I one
pass operation
.IR only ,
in order to display the associated
.B groff
version information, before exiting.
.
.
.\" --------------------------------------------------------------------
.
.SH ENVIRONMENT
The following environment variables may be set, and exported,
to modify the behaviour of
.BR pdfroff .
.
.TP
.B PDFROFF_COLLATE
Specifies the program to be used
for collation of the finshed PDF document.
.IP
This collation step may be required to move
.I tables of contents
to the start of the finished PDF document,
when formatting with traditional macro packages,
which print them at the end.
However,
users should not normally need to specify
.BR \%PDFROFF_COLLATE ,
(and indeed,
are not encouraged to do so).
If unspecified,
.B pdfroff
uses
.BR sed (@MAN1EXT@)
by default,
which normally suffices.
.IP
If
.B \%PDFROFF_COLLATE
.I is
specified,
then it must act as a filter,
accepting a list of file name arguments,
and write its output to the
.I stdout
stream,
whence it is piped to the
.BR \%PDFROFF_POSTPROCESSOR_COMMAND ,
to produce the finished PDF output.
.IP
When specifying
.BR \%PDFROFF_COLLATE ,
it is normally necessary to also specify
.BR \%PDFROFF_KILL_NULL_PAGES .
.IP
.B \%PDFROFF_COLLATE
is ignored,
if
.B pdfroff
is invoked with the
.I \%\-\-no\-kill\-null\-pages
option.
.
.TP
.B PDFROFF_KILL_NULL_PAGES
Specifies options to be passed to the
.B \%PDFROFF_COLLATE
program.
.IP
It should not normally be necessary to specify
.BR \%PDFROFF_KILL_NULL_PAGES .
The internal default is a
.BR sed (@MAN1EXT@)
script,
which is intended to remove completely blank pages
from the collated output stream,
and which should be appropriate in most applications of
.BR pdfroff .
However,
if any alternative to
.BR sed (@MAN1EXT@)
is specified for
.BR \%PDFROFF_COLLATE ,
then it is likely that a corresponding alternative specification for
.B \%PDFROFF_KILL_NULL_PAGES
is required.
.IP
As in the case of
.BR \%PDFROFF_COLLATE ,
.B \%PDFROFF_KILL_NULL_PAGES
is ignored,
if
.B pdfroff
is invoked with the
.I \%\-\-no\-kill\-null\-pages
option.
.
.TP
.B PDFROFF_POSTPROCESSOR_COMMAND
Specifies the command to be used for the final document conversion
from PostScript intermediate output to PDF.
It must behave as a filter,
writing its output to the
.I stdout
stream,
and must accept an arbitrary number of
.I files .\|.\|.\&
arguments,
with the special case of
.I \-
representing the
.I stdin
stream.
.IP
If unspecified,
.B \%PDFROFF_POSTPROCESSOR_COMMAND
defaults to
.
.RS 2
.IP
.I
.ad l
.NH gs \-dBATCH \-dQUIET \-dNOPAUSE \-sDEVICE=pdfwrite \-sOutputFile=\-
.ad
.RE
.
.TP
.B GROFF_TMPDIR
Identifies the directory in which
.B pdfroff
should create temporary files.
If
.B \%GROFF_TMPDIR
is
.I not
specified, then the variables
.BR TMPDIR ,
.B TMP
and
.B TEMP
are considered in turn, as possible temporary file repositories.
If none of these are set, then temporary files are created
in the current directory.
.
.TP
.B GROFF_GHOSTSCRIPT_INTERPRETER
Specifies the program to be invoked, when
.B pdfroff
converts
.B groff
PostScript output to PDF.
If
.B \%PDFROFF_POSTPROCESSOR_COMMAND
is specified,
then the command name it specifies is
.I implicitly
assigned to
.BR \%GROFF_GHOSTSCRIPT_INTERPRETER ,
overriding any explicit setting specified in the environment.
If
.B \%GROFF_GHOSTSCRIPT_INTERPRETER
is not specified, then
.B pdfroff
searches the process
.BR PATH ,
looking for a program with any of the well known names
for the GhostScript interpreter;
if no GhostScript interpreter can be found,
.B pdfroff
aborts.
.
.TP
.B GROFF_AWK_INTERPRETER
Specifies the program to be invoked, when
.B pdfroff

src/cmd/groff/contrib/pdfmark/pdfroff.sh
#! /bin/sh
# ------------------------------------------------------------------------------
#
# Function: Format PDF Output from groff Markup
#
# Copyright (C) 2005, 2006, Free Software Foundation, Inc.
# Written by Keith Marshall (keith.d.marshall@ntlworld.com)
# 
# This file is part of groff.
# 
# groff is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free
# Software Foundation; either version 2, or (at your option) any later
# version.
# 
# groff is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
# 
# You should have received a copy of the GNU General Public License along
# with groff; see the file COPYING.  If not, write to the Free Software
# Foundation, 51 Franklin St - Fifth Floor, Boston, MA 02110-1301, USA.
#
# ------------------------------------------------------------------------------
#
# Set up an identifier for the NULL device.
# In most cases "/dev/null" will be correct, but some shells on
# MS-DOS/MS-Windows systems may require us to use "NUL".
#
  NULLDEV="/dev/null"
  test -c $NULLDEV || NULLDEV="NUL"
#
# Set up the command name to use in diagnostic messages.
# (We can't assume we have 'basename', so use the full path if required.
#  Also use the 'exec 2>...' workaround for a bug in Cygwin's 'ash').
#
  CMD=`exec 2>$NULLDEV; basename $0` || CMD=$0
#
# To ensure that prerequisite helper programs are available, and are
# executable, a [fairly] portable method of detecting such programs is
# provided by function `searchpath'.
#
  searchpath(){
  #
  # Usage:  searchpath progname path
  #
    IFS=${PATH_SEPARATOR-":"} prog=':'
    for dir in $2
    do
      for ext in '' '.exe'
      #
      # try `progname' with all well known extensions
      # (e.g. Win32 may require `progname.exe')
      #
      do
        try="$dir/$1$ext"
        test -f "$try" && test -x "$try" && prog="$try" && break
      done
      test "$prog" = ":" || break
    done
    echo "$prog"
  }
# @PATH_SEARCH_SETUP@
#
# If the system maps '/bin/sh' to some 'zsh' implementation,
# then we may need this hack, adapted from autoconf code.
#
  test x${ZSH_VERSION+"set"} = x"set" && NULLCMD=":" \
    && (emulate sh) >$NULLDEV 2>&1 && emulate sh
#
# We need both 'grep' and 'sed' programs, to parse script options,
# and we also need 'cat', to display help and some error messages,
# so ensure they are all installed, before we continue.
#
  CAT=`searchpath cat "$PATH"`
  GREP=`searchpath grep "$PATH"`
  SED=`searchpath sed "$PATH"`
#
# Another fundamental requirement is the 'groff' program itself;
# we MUST use a 'groff' program located in 'GROFF_BIN_DIR', if this
# is specified; if not, we will search 'GROFF_BIN_PATH', only falling
# back to a 'PATH' search, if neither of these is specified.
#
  if test -n "$GROFF_BIN_DIR"
  then
    GPATH=GROFF_BIN_DIR
    GROFF=`searchpath groff "$GROFF_BIN_DIR"`
#
  elif test -n "$GROFF_BIN_PATH"
  then
    GPATH=GROFF_BIN_PATH
    GROFF=`searchpath groff "$GROFF_BIN_PATH"`
#
  else
    GPATH=PATH
    GROFF=`searchpath groff "$PATH"`
  fi
#
# If one or more of these is missing, diagnose and bail out.
#
  NO='' NOPROG="$CMD: installation problem: cannot find program"
  test "$CAT" = ":" && echo >&2 "$NOPROG 'cat' in PATH" && NO="$NO 'cat'"
  test "$GREP" = ":" && echo >&2 "$NOPROG 'grep' in PATH" && NO="$NO 'grep'"
  test "$GROFF" = ":" && echo >&2 "$NOPROG 'groff' in $GPATH" && NO="$NO 'groff'"
  test "$SED" = ":" && echo >&2 "$NOPROG 'sed' in PATH" && NO="$NO 'sed'"
  if test -n "$NO"
  then
    set $NO
    test $# -gt 1 && NO="s" IS="are" || NO='' IS="is"
    while test $# -gt 0
    do
      test $# -gt 2 && NO="$NO $1,"
      test $# -eq 2 && NO="$NO $1 and" && shift
      test $# -lt 2 && NO="$NO $1"
      shift
    done
    $CAT >&2 <<-ETX

	*** FATAL INSTALLATION ERROR ***

	The program$NO $IS required by '$CMD',
	but cannot be found; '$CMD' is unable to continue.

	ETX
    exit 1
  fi
#
# Identify the postprocessor command, for writing PDF output.
# (May be forced, by defining PDFROFF_POSTPROCESSOR_COMMAND in the environment;
#  if this is not set, leave blank to use the built in default).
#
  if test -n "${PDFROFF_POSTPROCESSOR_COMMAND}"
  then
    GROFF_GHOSTSCRIPT_INTERPRETER=`set command ${PDFROFF_POSTPROCESSOR_COMMAND};
      echo $2`
  fi
#
# Set up temporary/intermediate file locations.
#
  WRKFILE=${GROFF_TMPDIR=${TMPDIR-${TMP-${TEMP-"."}}}}/pdf$$.tmp
#
  REFCOPY=${GROFF_TMPDIR}/pdf$$.cmp
  REFFILE=${GROFF_TMPDIR}/pdf$$.ref
#
  CS_DATA=""
  TC_DATA=${GROFF_TMPDIR}/pdf$$.tc
  BD_DATA=${GROFF_TMPDIR}/pdf$$.ps
#
# Set a trap, to delete temporary files on exit.
# (FIXME: may want to include other signals, in released version).
#
  trap "rm -f ${GROFF_TMPDIR}/pdf$$.*" 0
#
# Initialise 'groff' format control settings,
# to discriminate table of contents and document body formatting passes.
#
  TOC_FORMAT="-rPHASE=1"
  BODY_FORMAT="-rPHASE=2"
#
  LONGOPTS="
    help	reference-dictionary	no-reference-dictionary
    stylesheet	pdf-output		no-pdf-output
    version	report-progress		no-toc-relocation
    emit-ps	keep-temporary-files	no-kill-null-pages
    "
# Parse the command line, to identify 'pdfroff' specific options.
# Collect all other parameters into new argument and file lists,
# to be passed on to 'groff', enforcing the '-Tps' option.
#
  DIFF="" STREAM="" INPUT_FILES=""
  SHOW_VERSION="" GROFF_STYLE="$GROFF -Tps"
  while test $# -gt 0
  do
    case "$1" in
#
#     Long options must be processed locally ...
#
      --*)
#
#          First identify, matching any abbreviation to its full form.
#
           MATCH="" OPTNAME=`IFS==; set dummy $1; echo $2`
           for OPT in $LONGOPTS
           do
             MATCH="$MATCH"`echo --$OPT | $GREP "^$OPTNAME"`
           done
#
#          For options in the form --option=value
#          capture any specified value into $OPTARG.
#
	   OPTARG=`echo $1 | $SED -n s?"^${OPTNAME}="??p`
#
#          Perform case specific processing for matched option ...
#
           case "$MATCH" in

             --help)
               $CAT <<-ETX
		Usage: $CMD [-option ...] [--long-option ...] [file ...]

		Options:
		  -h
		  --help
		 	Display this usage summary, and exit.

		  -v
		  --version
		 	Display a version identification message and exit.

		  --report-progress
		  	Enable console messages, indicating the progress of the
		 	PDF document formatting process.

		  --emit-ps
		  	Emit PostScript output instead of PDF; this may be useful
		 	when the ultimate PDF output is to be generated by a more
		 	specialised postprocessor, (e.g. gpresent), rather than
		 	the default GhostScript PDF writer.

		  --pdf-output=name
		  	Write the PDF, (or PostScript), output stream to file
		 	'name'; if this option is unspecified, standard output
		 	is used for PDF, (or PostScript), output.

		  --no-pdf-output
		 	Suppress the generation of PDF, (or PostScript), output
		 	entirely; use this with the --reference-dictionary option,
		 	if processing a document stream to produce only a
		 	reference dictionary.

		  --no-reference-dictionary
		 	Suppress the generation of a '$CMD' reference dictionary
		 	for the PDF document.  Normally '$CMD' will create a
		 	reference dictionary, at the start of document processing;
		 	this option can accelerate processing, if it is known in
		 	advance, that no reference dictionary is required.

		  --reference-dictionary=name
		 	Save the document reference dictionary in file 'name'.
		 	If 'name' already exists, when processing commences, it
		 	will be used as the base case, from which the updated
		 	dictionary will be derived.  If this option is not used,
		 	then the reference dictionary, created during the normal
		 	execution of '$CMD', will be deleted on completion of
		 	document processing.

		  --stylesheet=name
		  	Use the file 'name' as a 'groff' style sheet, to control
		 	the appearance of the document's front cover section.  If
		 	this option is not specified, then no special formatting
		 	is applied, to create a front cover section.

		  --no-toc-relocation
		 	Suppress the multiple pass 'groff' processing, which is
		 	normally required to position the table of contents at the
		 	start of a PDF document.

		  --no-kill-null-pages
		 	Suppress the 'null page' elimination filter, which is used
		 	to remove the excess blank pages produced by the collation
		 	algorithm used for 'toc-relocation'.

		  --keep-temporary-files
		 	Suppress the normal clean up of temporary files, which is
		 	scheduled when 'pdfroff' completes.

		ETX
               exit 0
               ;;

             --version)
	       GROFF_STYLE="$GROFF_STYLE \"$1\""
               SHOW_VERSION="GNU pdfroff (groff) version @VERSION@"
               ;;

             --report-progress)
               SHOW_PROGRESS=echo
               ;;

             --keep-temporary-files)
               trap "" 0
               ;;

	     --emit-ps)
	       PDFROFF_POSTPROCESSOR_COMMAND="$CAT"
	       ;;

             --pdf-output)
	       PDF_OUTPUT="$OPTARG"
	       ;;

	     --no-pdf-output)
	       PDF_OUTPUT="$NULLDEV"
	       ;;

             --reference-dictionary)
               REFFILE="$OPTARG"
               ;;

             --no-reference-dictionary)
               AWK=":" DIFF=":" REFFILE="$NULLDEV" REFCOPY="$NULLDEV"
               ;;

             --stylesheet)
               STYLESHEET="$OPTARG" CS_DATA=${GROFF_TMPDIR}/pdf$$.cs
               ;;

	     --no-toc-relocation)
	       TC_DATA="" TOC_FORMAT="" BODY_FORMAT=""
	       ;;

	     --no-kill-null-pages)
	       PDFROFF_COLLATE="$CAT" PDFROFF_KILL_NULL_PAGES=""
	       ;;
#
#          any other non-null match must have matched more than one defined case,
#          so report the ambiguity, and bail out.
#
             --*)
               echo >&2 "$CMD: ambiguous abbreviation in option '$1'"
	       exit 1
               ;;
#
#          while no match at all simply represents an undefined case.
#
             *)
               echo >&2 "$CMD: unknown option '$1'"
	       exit 1
               ;;
           esac
           ;;
#
#     A solitary hyphen, as an argument, means "stream STDIN through groff",
#     while the "-i" option means "append STDIN stream to specified input files",
#     so set up a mechanism to achieve this, for ALL 'groff' passes.
#
      - | -i*)
	   STREAM="$CAT ${GROFF_TMPDIR}/pdf$$.in |"
	   test "$1" = "-" && INPUT_FILES="$INPUT_FILES $1" \
	     || GROFF_STYLE="$GROFF_STYLE $1" 
	   ;;
#
#     Those standard options which expect an argument, but are specified with
#     an intervening space, between flag and argument, must be reparsed, so we
#     can trap invalid use of '-T dev', or missing input files.
#
      -[dfFILmMnoPrTwW])
           OPTNAME="$1"
	   shift; set reparse "$OPTNAME$@"
	   ;;
#
#     Among standard options, '-Tdev' is treated as a special case.
#     '-Tps' is automatically enforced, so if specified, is silently ignored.
#
      -Tps) ;;
#
#     No other '-Tdev' option is permitted.
#
      -T*) echo >&2 "$CMD: option '$1' is incompatible with PDF output"
           exit 1
	   ;;
#
#     '-h' and '-v' options redirect to their equivalent long forms ...
#
      -h*) set redirect --help
           ;;
#
      -v*) shift; set redirect --version "$@"
           ;;
#
#     All other standard options are simply passed through to 'groff',
#     with no validation beforehand.
#
      -*)  GROFF_STYLE="$GROFF_STYLE \"$1\""
           ;;
#
#     All non-option arguments are considered as possible input file names,
#     and are passed on to 'groff', unaltered.
#
      *)   INPUT_FILES="$INPUT_FILES \"$1\""
           ;;
    esac
    shift
  done
#
# If the '-v' or '--version' option was specified,
# then we simply emulate the behaviour of 'groff', with this option,
# and quit.
#
  if test -n "$SHOW_VERSION"
  then
    echo >&2 "$SHOW_VERSION"
    echo >&2; eval $GROFF_STYLE $INPUT_FILES
    exit $?
  fi
#
# Establish how to invoke 'echo', suppressing the terminating newline.
# (Adapted from 'autoconf' code, as found in 'configure' scripts).
#
  case `echo "testing\c"; echo 1,2,3`,`echo -n testing; echo 1,2,3` in
    *c*,*-n*)  n=''   c=''   ;;
    *c*)       n='-n' c=''   ;;
    *)         n=''   c='\c' ;;
  esac
#
# If STDIN is specified among the input files,
# or if no input files are specified, then we need to capture STDIN,
# so we can replay it into each 'groff' processing pass.
#
  test -z "$INPUT_FILES" && STREAM="$CAT ${GROFF_TMPDIR}/pdf$$.in |"
  test -n "$STREAM" && $CAT > ${GROFF_TMPDIR}/pdf$$.in
#
# Unless reference resolution is explicitly suppressed,
# we initiate it by touching the cross reference dictionary file,
# and initialise the comparator, to kickstart the reference resolver loop.
#
  SAY=":"
  if test -z "$DIFF"
  then
    >> $REFFILE
    echo kickstart > $REFCOPY
    test x${SHOW_PROGRESS+"set"} = x"set" && SAY=echo
#
#   In order to correctly resolve 'pdfmark' references,
#   we need to have both the 'awk' and 'diff' programs available.
#
    NO=''
    if test -n "$GROFF_AWK_INTERPRETER"
    then
      AWK="$GROFF_AWK_INTERPRETER"
      test -f "$AWK" && test -x "$AWK" || AWK=":"
    else
      for prog in @GROFF_AWK_INTERPRETERS@
      do
	AWK=`searchpath $prog "$PATH"`
	test "$AWK" = ":" || break
      done
    fi
    DIFF=`searchpath diff "$PATH"`
    test "$AWK" = ":" && echo >&2 "$NOPROG 'awk' in PATH" && NO="$NO 'awk'"
    test "$DIFF" = ":" && echo >&2 "$NOPROG 'diff' in PATH" && NO="$NO 'diff'"
    if test -n "$NO"
    then
      set $NO
      SAY=":" AWK=":" DIFF=":"
      test $# -gt 1 && NO="s $1 and $2 are" || NO=" $1 is"
      $CAT >&2 <<-ETX

	*** WARNING ***

	The program$NO required, but cannot be found;
	consequently, '$CMD' is unable to resolve 'pdfmark' references.

	Document processing will continue, but no 'pdfmark' reference dictionary
	will be compiled; if any 'pdfmark' reference appears in the resulting PDF
	document, the formatting may not be correct.

	ETX
    fi
  fi
#
# Run the multi-pass 'pdfmark' reference resolver loop ...
#
  $SAY >&2 $n Resolving references ..$c
  until $DIFF $REFCOPY $REFFILE 1>$NULLDEV 2>&1
  do
#
#   until all references are resolved, to yield consistent values
#   in each of two consecutive passes, or until it seems that no consistent
#   resolution is achievable.
#
    $SAY >&2 $n .$c
    PASS_INDICATOR="${PASS_INDICATOR}."
    if test "$PASS_INDICATOR" = "...."
    then
#
#     More than three passes required indicates a probable inconsistency
#     in the source document; diagnose, and bail out.
#
      $SAY >&2 " failed"
      $CAT >&2 <<-ETX
	$CMD: unable to resolve references consistently after three passes
	$CMD: the source document may exhibit instability about the reference(s) ...
	ETX
#
#     Report the unresolved references, as a diff between the two pass files,
#     preferring 'unified' or 'context' diffs, when available
#
      DIFFOPT=''
      $DIFF -c0 $NULLDEV $NULLDEV 1>$NULLDEV 2>&1 && DIFFOPT='-c0'
      $DIFF -u0 $NULLDEV $NULLDEV 1>$NULLDEV 2>&1 && DIFFOPT='-u0'
      $DIFF >&2 $DIFFOPT $REFCOPY $REFFILE
      exit 1
    fi
#
#   Replace the comparison file copy from any previous pass,
#   with the most recently updated copy of the reference dictionary.
#   (Some versions of 'mv' may not support overwriting of an existing file,
#    so remove the old comparison file first).
#
    rm -f $REFCOPY
    mv $REFFILE $REFCOPY
#
#   Run 'groff' and 'awk', to identify reference marks in the document source,
#   filtering them into the reference dictionary; discard incomplete 'groff' output
#   at this stage.
#
    eval $STREAM $GROFF_STYLE -Z 1>$NULLDEV 2>$WRKFILE $REFCOPY $INPUT_FILES
    $AWK '/^gropdf-info:href/ {$1 = ".pdfhref D -N"; print}' $WRKFILE > $REFFILE
  done
  $SAY >&2 " done"
#
# To get to here ...
# We MUST have resolved all 'pdfmark' references, such that the content of the
# updated reference dictionary file EXACTLY matches the last saved copy.
#
# If PDF output has been suppressed, then there is nothing more to do.
#
  test "$PDF_OUTPUT" = "$NULLDEV" && exit 0
#
# We are now ready to start preparing the intermediate PostScript files,
# from which the PDF output will be compiled -- but before proceding further ...
# let's make sure we have a GhostScript interpreter to convert them!
#
  if test -n "$GROFF_GHOSTSCRIPT_INTERPRETER"
  then
    GS="$GROFF_GHOSTSCRIPT_INTERPRETER"
    test -f "$GS" && test -x "$GS" || GS=":"
  else
    for prog in @GROFF_GHOSTSCRIPT_INTERPRETERS@
    do
      GS=`searchpath $prog "$PATH"`
      test "$GS" = ":" || break
    done
  fi
#
# If we could not find a GhostScript interpreter, then we can do no more.
#
  if test "$GS" = ":"
  then
    echo >&2 "$CMD: installation problem: cannot find GhostScript interpreter"
    $CAT >&2 <<-ETX

	*** FATAL INSTALLATION ERROR ***

	'$CMD' requires a GhostScript interpreter to convert PostScript to PDF.
	Since you do not appear to have one installed, '$CMD' connot continue.

	ETX
    exit 1
  fi
#
# We now extend the local copy of the reference dictionary file,
# to create a full 'pdfmark' reference map for the document ...
#
  $AWK '/^grohtml-info/ {print ".pdfhref Z", $2, $3, $4}' $WRKFILE >> $REFCOPY
#
# Re-enable progress reporting, if necessary ...
# (Missing 'awk' or 'diff' may have disabled it, to avoid display
#  of spurious messages associated with reference resolution).
#
  test x${SHOW_PROGRESS+"set"} = x"set" && SAY=echo
#
# If a document cover style sheet is specified ...
# then we run a special formatting pass, to create a cover section file.
#
  if test -n "$STYLESHEET"
  then
    DOT='^\.[ 	]*'
    CS_MACRO=${CS_MACRO-"CS"} CE_MACRO=${CE_MACRO-"CE"}
    $SAY >&2 $n "Formatting document ... front cover section ..$c"
    CS_FILTER="$STREAM $SED -n '/${DOT}${CS_MACRO}/,/${DOT}${CE_MACRO}/p'"
    eval $CS_FILTER $INPUT_FILES | eval $GROFF_STYLE $STYLESHEET - > $CS_DATA
    $SAY >&2 ". done"
  fi
#
# If table of contents relocation is to be performed (it is, by default),
# then we run an extra 'groff' pass, to format a TOC intermediate file.
#
  if test -n "$TC_DATA"
  then
    $SAY >&2 $n "Formatting document ... table of contents ..$c"
    eval $STREAM $GROFF_STYLE $TOC_FORMAT $REFCOPY $INPUT_FILES > $TC_DATA
    $SAY >&2 ". done"
  fi
#
# In all cases, a final 'groff' pass is required, to format the document body.
#
  $SAY >&2 $n "Formatting document ... body section ..$c"
  eval $STREAM $GROFF_STYLE $BODY_FORMAT $REFCOPY $INPUT_FILES > $BD_DATA
  $SAY >&2 ". done"
#
# Finally ...
# Invoke GhostScript as a PDF writer, to bind all of the generated
# PostScript intermediate files into a single PDF output file.
#
  $SAY >&2 $n "Writing PDF output ..$c"
  if test -z "$PDFROFF_POSTPROCESSOR_COMMAND"
  then
    PDFROFF_POSTPROCESSOR_COMMAND="$GS -dQUIET -dBATCH -dNOPAUSE
      -sDEVICE=pdfwrite -sOutputFile="${PDF_OUTPUT-"-"}

  elif test -n "$PDF_OUTPUT"
  then
    exec > $PDF_OUTPUT
  fi
#
# (This 'sed' script is a hack, to eliminate redundant blank pages).
#
  ${PDFROFF_COLLATE-"$SED"} ${PDFROFF_KILL_NULL_PAGES-'
      /%%Page:/{
	N
	/%%BeginPageSetup/b again
      }
      b
    :again
      /%%EndPageSetup/b finish
      /%%Page:/{
	N
	b again
      }
      b
    :finish
      N
      /^%%Page:.*\n0 Cg EP$/d
    '} $TC_DATA $BD_DATA | $PDFROFF_POSTPROCESSOR_COMMAND $CS_DATA -
  $SAY >&2 ". done"
#
# ------------------------------------------------------------------------------
# $RCSfile: pdfroff.sh,v $ $Revision: 1.13 $: end of file
